### PR TITLE
Ensure locking works with eggs

### LINKED ISSUE #1849: Resolver should ignore eggs when fetching egg_info
Recently a months old `Pipfile` has begun to throw errors when attempting to `pipenv install|update|lock`.

Our `Pipfile` references the `RandomWords==0.2.0` package, as of today, install/update/lock fails with:

```
hon3.6/site-packages/pipenv/patched/piptools/repositories/pypi.py", line 174, in get_dependencies
    legacy_results = self.get_legacy_dependencies(ireq)
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/piptools/repositories/pypi.py", line 222, in get_legacy_dependencies
    result = reqset._prepare_file(self.finder, ireq, ignore_requires_python=True)
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/notpip/req/req_set.py", line 644, in _prepare_file
    abstract_dist.prep_for_dist()
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/notpip/req/req_set.py", line 134, in prep_for_dist
    self.req_to_install.run_egg_info()
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/pip9/req/req_install.py", line 439, in run_egg_info
    command_desc='python setup.py egg_info')
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/pip9/utils/__init__.py", line 707, in call_subprocess
    % (command_desc, proc.returncode, cwd))
pip9.exceptions.InstallationError: Command "python setup.py egg_info" failed with error code 1 in /var/folders/5g/27jcjshj6sj014r0b6m49n4c0000gp/T/tmp1rmfvvbuild/randomwords/
```

I fired up a blank macOS installation and took the following steps to reproduce:

```
brew install python@2 python

pip3 install pipenv

mkdir ~/scratch ; cd ~/scratch

pipenv --two install --verbose "RandomWords==0.2.*"
```

I can, however, just do `pipenv shell`, and then `pip install RandomWords` within the subshell and it's fine.

I have also tried the same with installing `pipenv` under `python@2`'s pip and it also fails with the same output:

**Debug Output**

```
                          ROUND 1
Current constraints:
  randomwords==0.2.*

Finding the best candidates:
  found candidate randomwords==0.2.1 (constraint was ==0.2.*)

Finding secondary dependencies:
  randomwords==0.2.1 not in cache, need to check index

INFO:notpip.req.req_set:Collecting randomwords==0.2.1
INFO:pip9.download:File was already downloaded /Users/derekmortimer/Library/Caches/pipenv/pkgs/RandomWords-0.2.1-py2.7.egg
INFO:pip9.utils:Complete output from command python setup.py egg_info:
INFO:pip9.utils:Traceback (most recent call last):
  File "<string>", line 1, in <module>
IOError: [Errno 2] No such file or directory: '/var/folders/5g/27jcjshj6sj014r0b6m49n4c0000gp/T/tmpHExRE4build/randomwords/setup.py'

----------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/pipenv/resolver.py", line 82, in <module>
    main()
  File "/usr/local/lib/python3.6/site-packages/pipenv/resolver.py", line 71, in main
    clear=do_clear,
  File "/usr/local/lib/python3.6/site-packages/pipenv/resolver.py", line 63, in resolve
    verbose=verbose,
  File "/usr/local/lib/python3.6/site-packages/pipenv/utils.py", line 425, in resolve_deps
    pre,
  File "/usr/local/lib/python3.6/site-packages/pipenv/utils.py", line 336, in actually_resolve_reps
    resolved_tree.update(resolver.resolve(max_rounds=PIPENV_MAX_ROUNDS))
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/piptools/resolver.py", line 102, in resolve
    has_changed, best_matches = self._resolve_one_round()
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/piptools/resolver.py", line 200, in _resolve_one_round
    for dep in self._iter_dependencies(best_match):
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/piptools/resolver.py", line 297, in _iter_dependencies
    dependencies = self.repository.get_dependencies(ireq)
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/piptools/repositories/pypi.py", line 174, in get_dependencies
    legacy_results = self.get_legacy_dependencies(ireq)
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/piptools/repositories/pypi.py", line 222, in get_legacy_dependencies
    result = reqset._prepare_file(self.finder, ireq, ignore_requires_python=True)
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/notpip/req/req_set.py", line 644, in _prepare_file
    abstract_dist.prep_for_dist()
  File "/usr/local/lib/python3.6/site-packages/pipenv/patched/notpip/req/req_set.py", line 134, in prep_for_dist
    self.req_to_install.run_egg_info()
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/pip9/req/req_install.py", line 439, in run_egg_info
    command_desc='python setup.py egg_info')
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/pip9/utils/__init__.py", line 707, in call_subprocess
    % (command_desc, proc.returncode, cwd))
pip9.exceptions.InstallationError: Command "python setup.py egg_info" failed with error code 1 in /var/folders/5g/27jcjshj6sj014r0b6m49n4c0000gp/T/tmpHExRE4build/randomwords/
```

----

**pipenv.help output**

`> python3 -m pipenv.help output`

<details><summary>$ python -m pipenv.help output</summary>

Pipenv version: `'11.9.0'`

Pipenv location: `'/usr/local/lib/python3.6/site-packages/pipenv'`

Python location: `'/usr/local/opt/python/bin/python3.6'`

Other Python installations in `PATH`:

  - `2.7`: `/usr/local/bin/python2.7`
  - `2.7`: `/usr/local/bin/python2.7`
  - `2.7`: `/usr/bin/python2.7`
  - `3.6`: `/usr/local/bin/python3.6m`
  - `3.6`: `/usr/local/bin/python3.6`

  - `2.7.14`: `/usr/local/bin/python`
  - `2.7.10`: `/usr/bin/python`
  - `2.7.14`: `/usr/local/bin/python2`
  - `3.6.4`: `/usr/local/bin/python3`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.6.4',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '17.4.0',
 'platform_system': 'Darwin',
 'platform_version': 'Darwin Kernel Version 17.4.0: Sun Dec 17 09:19:54 PST '
                     '2017; root:xnu-4570.41.2~1/RELEASE_X86_64',
 'python_full_version': '3.6.4',
 'python_version': '3.6',
 'sys_platform': 'darwin'}
```

System environment variables:

  - `TERM_SESSION_ID`
  - `SSH_AUTH_SOCK`
  - `Apple_PubSub_Socket_Render`
  - `COLORFGBG`
  - `ITERM_PROFILE`
  - `XPC_FLAGS`
  - `LANG`
  - `PWD`
  - `SHELL`
  - `SECURITYSESSIONID`
  - `TERM_PROGRAM_VERSION`
  - `TERM_PROGRAM`
  - `PATH`
  - `DISPLAY`
  - `COLORTERM`
  - `TERM`
  - `HOME`
  - `TMPDIR`
  - `USER`
  - `XPC_SERVICE_NAME`
  - `LOGNAME`
  - `__CF_USER_TEXT_ENCODING`
  - `ITERM_SESSION_ID`
  - `SHLVL`
  - `OLDPWD`
  - `PAGER`
  - `LESS`
  - `LC_CTYPE`
  - `LSCOLORS`
  - `VIRTUAL_ENV_DISABLE_PROMPT`
  - `ZNT_REPO_DIR`
  - `ZNT_CONFIG_DIR`
  - `MINIMAL_PWD_CHAR_LEN`
  - `GOPATH`
  - `CHTF_CURRENT`
  - `CHTF_CURRENT_TERRAFORM_VERSION`
  - `EDITOR`
  - `HOMEBREW_GITHUB_API_TOKEN`
  - `SBT_OPTS`
  - `JDK`
  - `JAVA_HOME`
  - `SSH_PUBLIC_KEY`
  - `VAGRANT_CLOUD_TOKEN`
  - `_`
  - `PYTHONDONTWRITEBYTECODE`
  - `PIP_PYTHON_PATH`
  - `PYTHONUNBUFFERED`

Pipenv–specific environment variables:

Debug–specific environment variables:

  - `PATH`: `/usr/local/Caskroom/terraform-0.10.7/0.10.7:/usr/local/opt/go/libexec/bin:/Users/derekmortimer/.go/bin:/Users/derekmortimer/.bf-scripts:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin`
  - `SHELL`: `/usr/local/bin/zsh`
  - `EDITOR`: `/usr/local/bin/code -w`
  - `LANG`: `en_GB.UTF-8`
  - `PWD`: `/Users/derekmortimer/Work/sandbox`

---------------------------

Contents of `Pipfile` ('/Users/derekmortimer/Work/sandbox/Pipfile'):

```toml
[[source]]
url = "https://pypi.python.org/simple"
verify_ssl = true
name = "pypi"

[packages]
randomwords = "==0.2.*"

[dev-packages]

[requires]
python_version = "2.7"

```

</details>

### PR DESCRIPTION
**Pull request by uranusjr**
Patch notpip to recognise .egg formats on PyPI. Close #1849, ref #1850.

Leaving the PyPI mocking part to @techalchemy :x

### CODE REVIEW

File: pipenv/patched/notpip/req/req_set.py [hunk 1]
```diff
@@ -97,8 +97,11 @@ def make_abstract_dist(req_to_install):
     """
     if req_to_install.editable:
         return IsSDist(req_to_install)
-    elif req_to_install.link and req_to_install.link.is_wheel:
-        return IsWheel(req_to_install)
+    elif req_to_install.link:
+        if req_to_install.link.is_wheel:
+            return IsWheel(req_to_install)
+        elif req_to_install.link.ext == '.egg':
+            return IsEgg(req_to_install)
```
Comment (techalchemy) [inline pipenv/patched/notpip/req/req_set.py]: I _think_ we can avoid the `IsEgg(DistAbstraction)` below if we just say:

```
elif req_to_install.link.ext == '.egg':
    return req_to_install.get_dist()
```
? or some variant on that

`InstallRequirement` objects can already do this for eggs, I think

**Comment by techalchemy** [comment]
Lets see if @vphilippon has any input at a glance as to why the resolver is currently prioritizing eggs over sdists.  Relevant hacks are probably [here](https://github.com/pypa/pipenv/blob/e4bdc0fb5bc22faa81841b173bd844d9b6dc4778/pipenv/patched/piptools/resolver.py#L313) and somewhere starting around [here](https://github.com/pypa/pipenv/blame/e4bdc0fb5bc22faa81841b173bd844d9b6dc4778/pipenv/patched/piptools/repositories/pypi.py#L179)

Not sure how useful  that is, but thanks

**Comment by vphilippon** [comment]
What I can tell is that the egg inclusion is due to the hack right here: https://github.com/pypa/pipenv/blob/master/pipenv/patched/notpip/index.py#L645

`ignore_compatibility=True` will skip the supported extensions check. 

Note that the installation already excludes the .egg, so its only the lock/resolver part that includes them right now.

**Comment by techalchemy** [comment]
@uranusjr I think this will actually fix _everything_ with this change, without the other branch (for the most part) -- so we can drop all the skips and whatnot from there and just bring over the improvements

**Comment by uranusjr** [comment]
This is much better than my hack 💮

**Comment by techalchemy** [comment]
@uranusjr what else do we need to get everything to pass

**Pull request merged**